mitiq/interface/mitiq_pennylane/tests/test_conversions_pennylane.py

```diff
@@ -0,0 +1,69 @@
+# Copyright (C) 2021 Unitary Fund
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import cirq
+import pennylane as qml
+from mitiq.interface.mitiq_pennylane import from_pennylane, to_pennylane
+from mitiq.utils import _equal
+
+
+def test_from_pennylane_with_expvals():
+    """Tests converting Pennylane expectation values."""
+    # <X>
+    with qml.tape.QuantumTape() as tape:
+        qml.expval(qml.PauliX(wires=[0]))
+
+    circuit = from_pennylane(tape)
+    q = cirq.LineQubit(0)
+    correct = cirq.Circuit(cirq.H.on(q), cirq.measure(q))
+    assert _equal(circuit, correct, require_qubit_equality=False)
+
+    # <Z>
+    with qml.tape.QuantumTape() as tape:
+        qml.expval(qml.PauliZ(wires=[0]))
+
+    circuit = from_pennylane(tape)
+    correct = cirq.Circuit(cirq.measure(q))
+    assert _equal(circuit, correct, require_qubit_equality=False)
+
+
+def test_from_pennylane():
+    with qml.tape.QuantumTape() as tape:
+            qml.CNOT(wires=[0, "a"])
+
+    circuit = from_pennylane(tape)
+    correct = cirq.Circuit(cirq.CNOT(*cirq.LineQubit.range(2)))
+
+    # TODO: from_pennylane adds measurements even if there is not measurements
+    #  in the tape. This is because tape.to_openqasm(...) measures all qubits
+    #  even if there are no measurements in the tape.
+    circuit = circuit[:1]  # Temp patch: Manually remove measurements.
+
+    assert _equal(circuit, correct, require_qubit_equality=False)
+
+
+def test_to_from_pennylane():
+    circuit = cirq.testing.random_circuit(
+        qubits=4, n_moments=2, op_density=1, random_state=1
+    )
+
+    converted = from_pennylane(to_pennylane(circuit))
+
+    print(circuit)
+    print(converted)
+
+    assert cirq.testing.assert_allclose_up_to_global_phase(
+        cirq.unitary(converted), cirq.unitary(circuit), atol=1e-7,
+    )
```

Comment:
Round-trip conversions not preserving qubit order. Output of this test:

```text
      ┌──┐
0: ─────×────────
        │
1: ────×┼────H───
       ││
2: ────×┼────@───
        │    │
3: ─────×────@───
      └──┘
            ┌─────────┐
q_0: ───×────H────────────M('c_0')───
        │
q_1: ───×────@────────────M('c_1')───
             │
q_2: ───×────┼M('c_2')───────────────
        │    │
q_3: ───×────@────────────M('c_3')───
            └─────────┘
```

Reply:
Hey @trbromley is it possible to reorder the wires of a quantum tape? (Long story motivating this question below.)

1. Input Cirq `circuit`:

```
      ┌──┐
0: ─────×────────
        │
1: ────×┼────H───
       ││
2: ────×┼────@───
        │    │
3: ─────×────@───
      └──┘
```

2. Resulting `tape` from `tape = to_pennylane(circuit)`:

```
 1: ──╭SWAP───H──┤  
 2: ──╰SWAP──╭C──┤  
 0: ──╭SWAP──│───┤  
 3: ──╰SWAP──╰Z──┤  
```

  - Note this is the same as above with permuted qubit order. Namely (0, 1, 2, 3) -> (1, 2, 0, 3).

3. `print(tape.to_openqasm(rotations=True))`

```
OPENQASM 2.0;
include "qelib1.inc";
qreg q[4];
creg c[4];
swap q[0],q[1];
h q[0];
swap q[2],q[3];
cz q[1],q[3];
measure q[0] -> c[0];
measure q[1] -> c[1];
measure q[2] -> c[2];
measure q[3] -> c[3];
```

  - Note the indices of the qubits in the QASM program do not correspond to the indices of the wires in the tape.
  - (Aside from main point about qubit order, but there are also added measurements that are not in the tape.)

4. When converting back to Cirq from this QASM, the qubit order is the same as in the QASM program, so is permuted:

```
            ┌─────────┐
q_0: ───×────H────────────M('c_0')───
        │
q_1: ───×────@────────────M('c_1')───
             │
q_2: ───×────┼M('c_2')───────────────
        │    │
q_3: ───×────@────────────M('c_3')───
            └─────────┘
```

This can be fixed by reordering the wires in step 2, hence the question.

Reply:
Sounds good, thanks! This is currently the main blocker for me. I'm fine with this suggested fix and the requirement that the tape's wires must be sortable. (Which isn't the case with wires labeled [0, "a"], for example.)

Reply:
Thanks for adding in the fix with pre-sorting the wires! Regarding the [follow up comment](https://github.com/PennyLaneAI/pennylane/issues/1500#issuecomment-893763852) on the PL-issue, is that a blocker for this PR? :thinking: